src/sync.ts

```diff
@@ -637,149 +750,44 @@ export class SyncApi {
-                // sync data to finish processing before we process the results of this one.
+            if (savedSyncToken) {
                 debuglog("Sending first sync request...");
-                this.currentSyncRequest = this.doSyncRequest({ filterId }, savedSyncToken);
+            } else {
+                debuglog("Sending initial sync request...");
+                const initialFilter = this.buildDefaultFilter();
+                initialFilter.setDefinition(filter.getDefinition());
+                initialFilter.setTimelineLimit(this.opts.initialSyncLimit);
+                // Use an inline filter, no point uploading it for a single usage
+                firstSyncFilter = JSON.stringify(initialFilter.getDefinition());
             }
 
-            // Now wait for the saved sync to finish...
-            debuglog("Waiting for saved sync before starting sync processing...");
-            await savedSyncPromise;
-            this.doSync({ filterId });
-        };
-
-        if (client.isGuest()) {
-            // no push rules for guests, no access to POST filter for guests.
-            this.doSync({});
-        } else {
-            // Pull the saved sync token out first, before the worker starts sending
-            // all the sync data which could take a while. This will let us send our
-            // first incremental sync request before we've processed our saved data.
-            debuglog("Getting saved sync token...");
-            savedSyncPromise = client.store.getSavedSyncToken().then((tok) => {
-                debuglog("Got saved sync token");
-                savedSyncToken = tok;
-                debuglog("Getting saved sync...");
-                return client.store.getSavedSync();
-            }).then((savedSync) => {
-                debuglog(`Got reply from saved sync, exists? ${!!savedSync}`);
-                if (savedSync) {
-                    return this.syncFromCache(savedSync);
-                }
-            }).catch(err => {
-                logger.error("Getting saved sync failed", err);
-            });
-            // Now start the first incremental sync request: this can also
-            // take a while so if we set it going now, we can wait for it
-            // to finish while we process our saved sync data.
-            getPushRules();
+            // Send this first sync request here so we can then wait for the saved
+            // sync data to finish processing before we process the results of this one.
+            this.currentSyncRequest = this.doSyncRequest({ filter: firstSyncFilter }, savedSyncToken);
         }
+
+        // Now wait for the saved sync to finish...
+        debuglog("Waiting for saved sync before starting sync processing...");
+        await this.savedSyncPromise;
+        // process the first sync request and continue syncing with the normal filterId
+        return this.doSync({ filter: filterId });
```

Comment:
Will this work as intended? If this filter has a timeline limit > 1 and you do an incremental sync where you previously only got 1 event, will that cause it to pull in more?

Reply:
Tests fine:

```bash
t3chguy@Michael-t3chguy-MBP ~> curl 'https://matrix-client.matrix.org/_matrix/client/r0/sync?filter=%7B%22room%22%3A%7B%22state%22%3A%7B%22lazy_load_members%22%3Atrue%7D%2C%22timeline%22%3A%7B%22limit%22%3A1%7D%7D%7D&timeout=0&_cacheBuster=1661267868536' \
                                     -H 'authority: matrix-client.matrix.org' \
                                     -H 'accept: application/json' \
                                     -H 'accept-language: en-GB,en;q=0.9' \
                                     -H 'authorization: Bearer XXX' \
                                     -H 'origin: http://localhost:8080' \
                                     -H 'sec-ch-ua: "Chromium";v="104", " Not A;Brand";v="99", "Google Chrome";v="104"' \
                                     -H 'sec-ch-ua-mobile: ?0' \
                                     -H 'sec-ch-ua-platform: "macOS"' \
                                     -H 'sec-fetch-dest: empty' \
                                     -H 'sec-fetch-mode: cors' \
                                     -H 'sec-fetch-site: cross-site' \
                                     -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36' \
                                     --compressed | jq -r '.next_batch, .rooms.join."!phxFokrQGfZBcMPPln:riot.ovh".timeline'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7834    0  7834    0     0   112k      0 --:--:-- --:--:-- --:--:--  134k
m3208786589~3.3208786609~2.3208786609~1.3208786609_757284974_610383_1545249272_1559480105_3502853_564923765_5019202378_0
{
  "events": [
    {
      "content": {
        "avatar_url": "mxc://matrix.org/uVZUPCuxefrZSebLyjpvOvRi",
        "displayname": "webdevguru_test16",
        "membership": "join"
      },
      "origin_server_ts": 1661267896354,
      "sender": "@webdevguru_test16:matrix.org",
      "state_key": "@webdevguru_test16:matrix.org",
      "type": "m.room.member",
      "unsigned": {
        "replaces_state": "$gxY-CsSYo0iJLIrmbZbA0WVxd0RJzhiH8waXnWKA2dw",
        "prev_content": {
          "avatar_url": "mxc://matrix.org/uVZUPCuxefrZSebLyjpvOvRi",
          "displayname": "webdevguru_test16",
          "membership": "invite"
        },
        "prev_sender": "@x:riot.ovh",
        "age": 302576
      },
      "event_id": "$gJZnU0UU0c78HoUlaz6ps9DYh-oQc999rzhN6G8fbwg"
    }
  ],
  "prev_batch": "t91-3208776565_757284974_610383_1545249272_1559480105_3502853_564923765_5019202378_0",
  "limited": true
}
t3chguy@Michael-t3chguy-MBP ~> curl 'https://matrix-client.matrix.org/_matrix/client/r0/sync?filter=%7B%22room%22%3A%7B%22state%22%3A%7B%22lazy_load_members%22%3Atrue%7D%2C%22timeline%22%3A%7B%22limit%22%3A20%7D%7D%7D&timeout=0&_cacheBuster=1661267868536&since=m3208786589~3.3208786609~2.3208786609~1.3208786609_757284974_610383_1545249272_1559480105_3502853_564923765_5019202378_0' \
                                     -H 'authority: matrix-client.matrix.org' \
                                     -H 'accept: application/json' \
                                     -H 'accept-language: en-GB,en;q=0.9' \
                                     -H 'authorization: Bearer XXX' \
                                     -H 'origin: http://localhost:8080' \
                                     -H 'sec-ch-ua: "Chromium";v="104", " Not A;Brand";v="99", "Google Chrome";v="104"' \
                                     -H 'sec-ch-ua-mobile: ?0' \
                                     -H 'sec-ch-ua-platform: "macOS"' \
                                     -H 'sec-fetch-dest: empty' \
                                     -H 'sec-fetch-mode: cors' \
                                     -H 'sec-fetch-site: cross-site' \
                                     -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36' \
                                     --compressed | jq -r '.next_batch, .rooms.join."!phxFokrQGfZBcMPPln:riot.ovh".timeline'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   829    0   829    0     0   4133      0 --:--:-- --:--:-- --:--:--  4631
s3208791343_757284974_617131_1545252597_1559484219_3502856_564925803_5019212484_0
{
  "events": [
    {
      "content": {
        "body": "a",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "a"
      },
      "origin_server_ts": 1661268210098,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$xw2hrC65b62rp3ABA-z03NHYLouIjrfBddCM0KMuBxs"
    },
    {
      "content": {
        "body": "b",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "b"
      },
      "origin_server_ts": 1661268210560,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$mSNLoSZYa9OST4m8yv1zMRgFL_bcQqDqfwNDUtY-JkQ"
    },
    {
      "content": {
        "body": "c",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "c"
      },
      "origin_server_ts": 1661268210897,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$5yeflSHZY-eCMU1F0fMPjpPutfaV6VEn1NiSlMeJLPk"
    },
    {
      "content": {
        "body": "d",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "d"
      },
      "origin_server_ts": 1661268211178,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$IbEEfy8AAVjdMlBqHnRBN1bsTZUk7gqVQcb7HQsvtXk"
    },
    {
      "content": {
        "body": "e",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "e"
      },
      "origin_server_ts": 1661268211503,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$EftWZvUsO0LoMpfaQu_yUccSYW3DPuhVAVKZDd_i2Q0"
    },
    {
      "content": {
        "body": "f",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "f"
      },
      "origin_server_ts": 1661268211779,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$TZkp_Gqoodg8A-oEj9YCVho21dtdYpWHellSBQdhC80"
    },
    {
      "content": {
        "body": "g",
        "msgtype": "m.text",
        "org.matrix.msc1767.text": "g"
      },
      "origin_server_ts": 1661268212443,
      "sender": "@x:riot.ovh",
      "type": "m.room.message",
      "unsigned": {},
      "event_id": "$yLzq2MSCq-8drNz-fAjlaxSIFWp6TX7Puk9AcLwPW-U"
    }
  ],
  "prev_batch": "s3208789296_757284974_617131_1545252597_1559484219_3502856_564925803_5019212484_0",
  "limited": false
}
```

Initial sync has limit=1 and returns a single event
Then I sent 7 events into the room
2nd sync has limit=20 and returns 7 events